### PR TITLE
fall back to eager on `NotImplementedError`

### LINKED ISSUE #107970: [Dynamo] Graph break on str arg to torch function
### 🐛 Describe the bug

Add this diff:
```
diff --git a/test/inductor/test_torchinductor_opinfo.py b/test/inductor/test_torchinductor_opinfo.py
index e886e52ce1e..fb9784e489b 100644
--- a/test/inductor/test_torchinductor_opinfo.py
+++ b/test/inductor/test_torchinductor_opinfo.py
@@ -206,7 +206,7 @@ inductor_expected_failures_single_sample["cpu"] = {
     "complex": {f16},
     "exponential": {f16},
     "geometric": {f16},
-    "linalg.eigh": {f32, f64},
+    # "linalg.eigh": {f32, f64},
     "linalg.eigvalsh": {f32, f64},
     "log_normal": {f16},
     "masked_scatter": {f16, f32, f64},
@@ -459,7 +459,7 @@ class TestInductorOpInfo(TestCase):
     @torch._inductor.config.patch(
         {"implicit_fallbacks": False, "triton.autotune_pointwise": False}
     )
-    @collection_decorator
+    # @collection_decorator
     def test_comprehensive(self, device, dtype, op):
         torch._dynamo.reset()
         with torch.no_grad():
@@ -589,6 +589,7 @@ class TestInductorOpInfo(TestCase):
                     )
 
         except Exception as e:
+            raise e
             known_failure = False
             if dtype in inductor_should_fail_with_exception[device_type].get(
                 op_name, set()
```

Run `python /home/eellison/pytorch/test/inductor/test_torchinductor_opinfo.py -k  eigh_cpu_float32`

>   File "/home/eellison/pytorch/torch/_dynamo/variables/torch.py", line 732, in call_function
    *proxy_args_kwargs(args, kwargs),
  File "/home/eellison/pytorch/torch/_dynamo/utils.py", line 497, in proxy_args_kwargs
    raise unimplemented(
  File "/home/eellison/pytorch/torch/_dynamo/exc.py", line 172, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: call_function args: TensorVariable() UserDefinedObjectVariable(str_)

This should be a simple fix. it seems like it's failing on converting the str type to value for running the op:

`torch.linalg.eigh(A, UPLO="U")`

### Versions

master

related: https://github.com/pytorch/pytorch/issues/105534

cc @mruberry @rgommers @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

### PR DESCRIPTION
**Pull request by ev-br**
Follow-up to https://github.com/pytorch/pytorch/pull/107710:

Help  dynamo fall back to eager when compiling unimplemented numpy constructs:

- arrays of strings
- (arg){min, max} for complex types
- various arguments typed as NotImplemented (`np.ones(4, order="F")` etc)
- numpy functions which torch._numpy does not implement

To test, run (we do not implement arrays of strings)

```
import torch
import numpy as np

@torch.compile(fullgraph=False)
def fn():
    return np.asarray(["L", "U"])
```

and observe it compiles with fullgraph=False and fails with fullgraph=True

Fixes https://github.com/pytorch/pytorch/issues/107970

cc @mruberry @rgommers @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @gujinghui @PenghuiCheng @jianyuh @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @anijain2305

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
Please  audit all the `NotImplementedError`s in the code. At the moment, we have a few others that fall outside of the `normalizer`, for example in `_ndarray`.

File: torch/_numpy/__init__.py [hunk 1]
```diff
@@ -26,3 +26,10 @@
 
 False_ = False
 True_ = True
+
+
+def __getattr__(name):
+    # is only called for names not found via a normal lookup (cf PEP 562)
+    # returning NotImplemented allows dynamo to fall back to eager,
+    # meaning the name will be searched on the NumPy namespace.
+    return NotImplemented
```
Comment (lezcano) [inline torch/_numpy/__init__.py:35]: Also, cool trick, but I don't think it's necessary. dynamo already knows how to fallback in this case.

**Comment by ev-br** [comment]
This PR should be failing CI left and right because of a syntax error (which I fixed locally bit did not push the fix). So can be used as a canary on a coalmine to check that torch_np tests actually run.

File: torch/_numpy/_normalizations.py [hunk 2]
```diff
@@ -193,23 +193,29 @@ def wrapped(*args, **kwds):
...
 
-            kwds = {
-                name: maybe_normalize(arg, params[name]) if name in params else arg
-                for name, arg in kwds.items()
-            }
+                kwds = {
+                    name: maybe_normalize(arg, params[name]) if name in params else arg
+                    for name, arg in kwds.items()
+                }
+            except (NotImplementedError, RuntimeError, ValueError):
```
[the remark this answers is not in the document]
Comment (ev-br) [reply torch/_numpy/_normalizations.py]: This try-except is gone in the revised version.

**Comment by ev-br** [comment]
Pushed an update to simplify the logic:

- in tnp.asarray, a failure to convert to tensor triggers NotImplementedError regardless of the reason
- `run_node`, convert catch NotImplementedError and convert it into  UnsupportedFakeTensorException

WDYT @lezcano?

File: torch/_dynamo/utils.py [hunk 3]
```diff
@@ -1406,6 +1409,10 @@ def run_node(tracer, node, args, kwargs, nnmodule):
         elif op == "placeholder":
             assert "example_value" in node.meta
             return node.meta["example_value"]
+    except Unsupported:
```
[the remark this answers is not in the document]
Comment (ev-br) [reply torch/_dynamo/utils.py]: The updated version catches `NotImplementedError`, so the logic is basically: graph break on NotImplementedError, error out on all other exceptions.
And this seems to be the place which dispatches various call_function, call_method and other operations.

File: torch/_numpy/_reductions_impl.py [hunk 4]
```diff
@@ -71,6 +71,9 @@ def argmax(
     *,
     keepdims: KeepDims = False,
 ):
+    if a.is_complex():
+        raise NotImplementedError
```
Comment (lezcano) [inline torch/_numpy/_reductions_impl.py:75]: Raise an instance. Same below.

**Review by lezcano — COMMENTED** [review]
This looks much better. That being said, I'm tagging Ed see if it's alright with him, or there's a better approach. Let's wait for his review.

**Comment by lezcano** [comment]
@ev-br the solution seems to break other tests. I think there must be another way of doing things, this one looks too hacky. Tagging @voznesenskym and @anijain2305 for ideas

File: test/torch_np/numpy_tests/lib/test_function_base.py [hunk 5]
```diff
@@ -3434,7 +3434,6 @@ def test_basic(self):
         a = np.array([0.0444502, 0.141249, 0.0463301])
         assert_equal(a[-1], np.median(a))
 
-    @pytest.mark.xfail(reason="median: scalar output vs 0-dim")
```
Comment (lezcano) [inline test/torch_np/numpy_tests/lib/test_function_base.py:3437]: Why did all these tests use to fail and now they pass?

**Review by ezyang — APPROVED** [review]
Seems fine. You weren't relying on NotImplemented sentinel to do multi dispatch for binary ops right?

**Comment by ezyang** [comment]
guess you gotta hose down some newly uncovered bugs now

**Comment by lezcano** [comment]
Nah, we added the `NotImplemented` as a workaround, but exceptions are better.

**Comment by ev-br** [comment]
The last commit xfails the sparse test from https://github.com/pytorch/pytorch/issues/108667

File: test/test_sparse.py [hunk 6]
```diff
@@ -591,6 +591,7 @@ def test_sparse_bool(self, device, dtype):
         b = a.to_sparse().to_dense()
         self.assertEqual(a, b)
 
+    @unittest.expectedFailure         # cf https://github.com/pytorch/pytorch/issues/108667
```
Comment (lezcano) [inline test/test_sparse.py:594]: Nope. We have a decorator that xfails if run with dynamo. That's what you want.

Comment (amjames) [reply test/test_sparse.py:594]: FWIW I did try this out locally and if that error were to change from `NotImplementedError` to something else, this all works fine.

Comment (ev-br) [reply test/test_sparse.py]: https://github.com/pytorch/pytorch/pull/107863/commits/8eabd14675241dd0a27b777ae8dd270b424c13ba `skipIfTorchDynamo`  as used elsewhere in this test suite

Comment (lezcano) [reply test/test_sparse.py]: @amjames yes, this is a small bug in TorchDynamo, that it should create a break properly and fallback to eager but it doesn't. If anyone from sparse wants to have a knock at this one it should be a good introductory PR.

**Review by lezcano — APPROVED** [review]
We should investigate at some point why the tests were passing before and not breaking now, but that can be done at a later time.

**Comment by atalman** [comment]
will be included in 2.2
